### PR TITLE
Move Code_id_or_symbol into Reg_width_things

### PR DESCRIPTION
**Pull request by mshinwell**
This allows `Code_id_or_symbol.t` to be represented as an integer, with the injection functions from `Code_id.t` and `Symbol.t` being the identity.

### CODE REVIEW

**Review by Gbury — APPROVED** [review]
I left a few comments, but it looks fine.

File: middle_end/flambda2/identifiers/reg_width_things.ml [hunk 1]
```diff
@@ -243,6 +245,42 @@ module Symbol_data = struct
...
+        @[<hov 1>(name@ %s)@]@ \
+        @[<hov 1>(linkage_name@ %a)@]@ \
+        )@]"
+      Compilation_unit.print compilation_unit
+      name
+      Linkage_name.print linkage_name
+
+  let hash { compilation_unit; name = _; linkage_name } =
+    Hashtbl.hash
+      (Compilation_unit.hash compilation_unit, Linkage_name.hash linkage_name)
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/identifiers/reg_width_things.ml:269]: Fixed (in the version I'm about to push)

File: middle_end/flambda2/identifiers/reg_width_things.ml [hunk 2]
```diff
@@ -243,6 +245,42 @@ module Symbol_data = struct
...
+      { compilation_unit = compilation_unit1;
+        name = _;
+        linkage_name = linkage_name1
+      }
+      { compilation_unit = compilation_unit2;
+        name = _;
+        linkage_name = linkage_name2
+      } =
+    Linkage_name.equal linkage_name1 linkage_name2
+    && Compilation_unit.equal compilation_unit1 compilation_unit2
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/identifiers/reg_width_things.ml:281]: No there are not.

File: middle_end/flambda2/identifiers/reg_width_things.ml [hunk 3]
```diff
@@ -725,8 +763,172 @@ module Simple = struct
...
+  type t = Table_by_int_id.Id.t
+
+  let create_code_id code_id = code_id
+
+  let create_symbol symbol = symbol
+
+  let pattern_match t ~code_id ~symbol =
+    if Table_by_int_id.Id.flags t = Code_id_data.flags
+    then code_id t
+    else symbol t
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/identifiers/reg_width_things.ml:874]: I've added a proper check + error

**Comment by mshinwell** [comment]
Guillaume's review will suffice for this.

**Pull request merged**